File: app/src/main/java/com/polok/eubmanagement/presentation/schedule/schedulelist/ScheduleListFragment.java
package com.polok.eubmanagement.presentation.schedule.schedulelist;

import android.os.Bundle;
import android.view.View;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import com.polok.eubmanagement.R;
import com.polok.eubmanagement.base.BaseFragment;
import com.polok.eubmanagement.base.BaseViewModel;
import com.polok.eubmanagement.databinding.FragmentScheduleListBinding;
import com.polok.eubmanagement.util.SharedPref;
import com.polok.eubmanagement.widget.PrimaryLoader;

public class ScheduleListFragment extends BaseFragment<FragmentScheduleListBinding> {
    @Override
    protected FragmentScheduleListBinding initViewBinding() {
        return FragmentScheduleListBinding.inflate(getLayoutInflater());
    }
    private ScheduleListViewModel viewModel;
    @Override
    protected BaseViewModel initViewModel() {
        return viewModel = new ViewModelProvider(this).get(ScheduleListViewModel.class);
    }
    @Override
    protected void initOnCreateView(Bundle savedInstanceState) {
        if (SharedPref.getUserProfile().getAdmin()) binding.addScheduleButton.setVisibility(View.VISIBLE);

        viewModel.fetchScheduleListFromFirebase();

        viewModel.getScheduleLiveData().observe(getViewLifecycleOwner(), scheduleList-> {
            if (scheduleList != null && !scheduleList.isEmpty()) {
                binding.scheduleRecycler.setAdapter(new ScheduleListAdapter(scheduleList));
            }
        });
        binding.addScheduleButton.setOnClickListener(view -> {
            viewModel.fireNavigateEvent(R.id.action_scheduleListFragment_to_scheduleAddFragment, null);
        });
    }
    @Override
    protected PrimaryLoader initPrimaryLoader() {return binding.primaryLoader;}
    @Override
    protected void onNavigateEvent(int id, Bundle bundle) {
        super.onNavigateEvent(id, bundle);
        Navigation.findNavController(binding.getRoot()).navigate(id, bundle);
    }
}

File: app/src/main/java/com/polok/eubmanagement/presentation/schedule/schedulelist/ScheduleListFragment.kt
package com.polok.eubmanagement.presentation.schedule.schedulelist

import android.os.Bundle
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import com.polok.eubmanagement.R
import com.polok.eubmanagement.base.BaseFragment
import com.polok.eubmanagement.base.BaseViewModel
import com.polok.eubmanagement.databinding.FragmentScheduleListBinding
import com.polok.eubmanagement.util.SharedPref
import com.polok.eubmanagement.util.makeVisible
import com.polok.eubmanagement.widget.PrimaryLoader

class ScheduleListFragment : BaseFragment<FragmentScheduleListBinding>(
    viewBindingFactory = FragmentScheduleListBinding::inflate
) {
    private val viewModel: ScheduleListViewModel by viewModels {
        ScheduleListViewModel.Factory()
    }
    private val adapter: ScheduleListAdapter by lazy {
        ScheduleListAdapter(
            onUpdateClickListener = {
                viewModel.navigateToScheduleUpdateFragment(it)
            }
        )
    }

    override fun initViewModel(): BaseViewModel = viewModel

    override fun initOnCreateView(savedInstanceState: Bundle?) {

        viewModel.fetchScheduleListFromFirebase()

        adapter.isAdmin = SharedPref.getUserProfile().admin
        if (adapter.isAdmin == true) binding.addButton.makeVisible()

        viewModel.scheduleLiveData.observe(viewLifecycleOwner) {
            if (it?.isNotEmpty() == true) {
                binding.scheduleRecycler.adapter = adapter.apply {
                    submitList(it)
                }
            }
        }

        binding.addButton.setOnClickListener {
            viewModel.fireNavigateEvent(
                R.id.action_scheduleListFragment_to_scheduleAddFragment, null
            )
        }
    }

    override fun initPrimaryLoader(): PrimaryLoader = binding.primaryLoader

    override fun onNavigateEvent(id: Int, bundle: Bundle?) {
        super.onNavigateEvent(id, bundle)
        findNavController().navigate(id, bundle)
    }
}